### PR TITLE
[libc] Replace -nostdlib++ flag when building with gcc and add placement new operator to HermeticTestUtils.cpp.

### PR DESCRIPTION
**Pull request by lntue**
`-nostdlib++` is a clang-only flag.  Replacing it with `-nostdlib` when building with gcc.

### CODE REVIEW

**Review by jhuber6 — COMMENTED** [review]
Does this work with `-nostdlib` on clang? If so it's probably better to be consistent.

**Comment by SchrodingerZhu** [comment]
https://github.com/llvm/llvm-project/blob/37ca558a9ed60cfff960f91dbb4dc30a70107601/libc/test/UnitTest/HermeticTestUtils.cpp#L26

Is this still true with `COMPILER_RT_BUILD_SCUDO_STANDALONE_WITH_LLVM_LIBC`?

**Comment by nickdesaulniers** [comment]
gcc-13 added support for `-nostdlib++`. Perhaps it's worth using `check_cxx_compiler_flag` to check for `-nostdlib++` support, then fall back to `-nostdlib` if not?

**Review by SchrodingerZhu — APPROVED** [review]
The change itself looks good to me. Just curious that if `COMPILER_RT_BUILD_SCUDO_STANDALONE_WITH_LLVM_LIBC` guarantees the compatibility.

File: libc/test/UnitTest/HermeticTestUtils.cpp [hunk 1]
```diff
@@ -104,6 +104,8 @@ void *__dso_handle = nullptr;
 
 } // extern "C"
 
+void *operator new(unsigned long size, void *ptr) { return ptr; }
```
Comment (michaelrj-google) [inline libc/test/UnitTest/HermeticTestUtils.cpp:107]: is this okay or should it call realloc?

**Review by michaelrj-google — COMMENTED** [review]
overall LGTM

File: libc/test/UnitTest/HermeticTestUtils.cpp (hunk 1 above)
Comment (lntue) [reply libc/test/UnitTest/HermeticTestUtils.cpp:107]: This is variation (9) in the list: https://en.cppreference.com/w/cpp/memory/new/operator_new
and it is said that: "The standard library implementation performs no action and returns `ptr` unmodified."

File: libc/cmake/modules/CheckCompilerFeatures.cmake [hunk 2]
```diff
@@ -57,3 +57,5 @@ foreach(feature IN LISTS ALL_COMPILER_FEATURES)
 endforeach()
 
 message(STATUS "Compiler features available: ${AVAILABLE_COMPILER_FEATURES}")
+
+check_cxx_compiler_flag("-nostdlib++" LIBC_COMPILER_SUPPORT_NOSTDLIBPP)
```
Comment (nickdesaulniers) [inline libc/cmake/modules/CheckCompilerFeatures.cmake:61]: will conflict with #78776

Comment (nickdesaulniers) [reply libc/cmake/modules/CheckCompilerFeatures.cmake:61]: Can you use a similar naming convention/prefix `LIBC_CC_SUPPORTS_`

Comment (lntue) [reply libc/cmake/modules/CheckCompilerFeatures.cmake]: Updated the flag name.  I can wait for you to merge #78776 first.

Comment (nickdesaulniers) [reply libc/cmake/modules/CheckCompilerFeatures.cmake]: > can wait for you to merge https://github.com/llvm/llvm-project/pull/78776 first.

done

**Review by nickdesaulniers — APPROVED** [review]

**Review by michaelrj-google — APPROVED** [review]

**Pull request merged**